Content.Server/MagicMirror/MagicMirrorSystem.cs

```diff
@@ -94,18 +101,23 @@ private void OnSelectSlotDoAfter(EntityUid uid, MagicMirrorComponent component,
                 return;
         }
 
-        _humanoid.SetMarkingId(component.Target.Value.Owner, category, args.Message.Slot, args.Message.Marking);
+        _humanoid.SetMarkingId(component.Target.Value, category, args.Message.Slot, args.Message.Marking);
 
-        UpdateInterface(uid, component.Target.Value.Owner, args.Message.Session);
+        UpdateInterface(uid, component.Target.Value, args.Message.Session);
     }
 
     private void OnTryMagicMirrorChangeColor(EntityUid uid, MagicMirrorComponent component, MagicMirrorChangeColorMessage message)
     {
         if (component.Target == null) return;
         if (message.Session.AttachedEntity == null) return;
 
+        var p1 = _transform.GetWorldPosition(component.Target.Value);
+        var p2 = _transform.GetWorldPosition(uid);
+        if (Vector2.Distance(p1, p2) > component.Range) return;
+
         var doAfter = new ChangeColorDoAfterEvent(message);
-        _doAfterSystem.TryStartDoAfter(new DoAfterArgs(EntityManager, message.Session.AttachedEntity.Value, component.ChangeSlotTime, doAfter, uid, target: component.Target.Value.Owner, used: uid)
+        _doAfterSystem.TryStartDoAfter(new DoAfterArgs(EntityManager, message.Session.AttachedEntity.Value, component.ChangeSlotTime, doAfter, uid, target: component.Target.Value, used: uid)
+
```

Comment:
I think at this point you're better off using the usual codepaths for this stuff instead of doing manual checks as this is going to be prone to all sorts of bugs, e.g. it never checks CanInteract. This means it should use a BUI or a verb.

Reply:
This should use InteractionSystem's InRangeUnobstructed and check both before & after the do-afters